proposals/0621-linear-constraints.rst

```diff
@@ -0,0 +1,1085 @@
+
+Final final nail: axioms of the form ``Traversable t %1 => (Functor
+t, Foldable t)`` aren't in the fragment that we know how to solve.
+
+Instance contexts
+^^^^^^^^^^^^^^^^^
+
+We do not specify a way, in this proposal, for instance contexts to be
+linear.
+
+That is the syntax
+
+::
+
+  instance (Foo a, Bar a) %1 => Baz a where {…}
+
+Is rejected. See alternatives for a potential specification.
+
+
+Wanted with Other Multiplicities
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Givens, by virtue of the syntax, are always either linear or
+unrestricted. However, wanted can, in principle, have different
+multiplicities.
+
+Let
+
+::
+
+   p :: Multiplicity
+   f :: A %p -> B
+   useC :: C %1 => A
+
+Then in
+
+::
+
+  f useC
+
+we have wanted ``C`` with multiplicity ``p`` (``p`` is a rigid
+variable). What do we do? We solve ``C`` as if it were an unrestricted
+wanted.
+
+Wanted with delayed multiplicities
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Consider
+
+::
+
+   f :: A %p -> (A %p -> B) -> B
+   useC :: C %1 => A
+
+Then in
+
+::
+
+   f useC
+
+we have wanted ``C`` whose multiplicity is a unification variable, the
+value of which will be determined by the context. What do we do? There
+are two cases:
+- There's an unrestricted given with head ``C`` *and no such linear given*, then
+  the unrestricted given can solve the wanted.
+- There is a linear given with head ``C``: we don't solve ``C`` until
+  ``p`` has been determined.
+
+Quantified Constraints
+^^^^^^^^^^^^^^^^^^^^^^
+
+Our syntax extension naturally extends constraints in types' context
+to support linear implications ``C %1 => D`` when
+``-XQuantifiedConstraint`` is on. While this is not described in the
+paper_, our solving algorithm is based on `Efficient resource
+management for linear logic proof search`_, where such higher-order
+givens are handled. The extension is unproblematic, it would
+presumably be more effort to prevent it than to support it.
+
+Therefore, when ``-XLinearTypes`` and ``-XLinearConstraints`` are both
+on, contexts can contain implications of the form ``C %1 => D``.
+
+Existential types proposal
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The `Existential types proposal`_, if they ever materialise (🤞), will
+make this proposal even better. Using linear constraints in APIs most
+often require returning constraints too. For that we've been using
+GADTs. This was our simple mutable array API:
+
+::
+
+  read  :: Read n %1 => MArray a n -> Int -> Read n /\ a
+
+  free :: (Read n, Write n) %1 => MArray a n -> ()
+
+  data c /\ a where
+    Box :: c %1 => a -> c /\ a
+
+  new :: Linearly %1 => Int -> NewMArray a
+
+  data NewMArray a where
+    NewMArray :: (Read n, Write n) %1 => MArray a n -> NewMArray a n
+
+With this API we can write functions such as
+
+::
+
+  read2AndDiscard ::  (Read n, Write n) %1 => MArray a n -> (Ur a, Ur a)
+  read2AndDiscard arr =
+       let !(Box x)  = read arr 0
+           !(Box y)  = read arr 1
+           !()       = free arr
+       (x, y)
```

Comment:
I don't understand if this is part of a proposed change to strict bindings. This wouldn't work in current GHC, would it? For example, in the below only `showItCase`  works, `showItLet` doesn't.

```haskell
data WithShow a where
  MkWithShow :: Show a => a -> WithShow a

showItCase :: WithShow a -> String
showItCase x = case x of MkWithShow a -> show a

showItLet :: WithShow a -> String
showItLet x = let !(MkWithShow a) = x in show a
```

Reply:
No, I just didn't realise it. I guess we'd done all of our examples with `case`.

I understand why it's so: these things desugar to a form that isn't conducive to manipulating existential types.

In fact, in the [linear let MR](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/10952), I'm changing the desugaring of this very case (strict, non-recursive, monomorphic pattern bind) to be a regular `case`. We could change the type checker to understand that such strict, non-recursive, monomorphic pattern binds can bind type variables and constraint evidence, but maybe it's a little fragile?